src/cargo/core/compiler/job_queue.rs

```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
             .take()
             .map(move |srv| srv.start(move |msg| drop(tx.send(Message::FixDiagnostic(msg)))));
 
-        // Use `crossbeam` to create a scope in which we can execute scoped
-        // threads. Note that this isn't currently required by Cargo but it was
-        // historically required. This is left in for now in case we need the
-        // `'a` ability for child threads in the near future, but if this
-        // comment has been sitting here for a long time feel free to refactor
-        // away crossbeam.
-        crossbeam_utils::thread::scope(|scope| self.drain_the_queue(cx, plan, scope, &helper))
+        crossbeam_utils::thread::scope(move |scope| state.drain_the_queue(cx, plan, scope, &helper))
             .expect("child threads shouldn't panic")
     }
+}
 
-    fn drain_the_queue(
+impl<'a, 'cfg> DrainState<'a, 'cfg> {
+    fn spawn_work_if_possible(
         &mut self,
         cx: &mut Context<'a, '_>,
+        jobserver_helper: &HelperThread,
+        scope: &Scope<'_>,
+        has_errored: bool,
+    ) -> CargoResult<()> {
+        // Dequeue as much work as we can, learning about everything
+        // possible that can run. Note that this is also the point where we
+        // start requesting job tokens. Each job after the first needs to
+        // request a token.
+        while let Some((unit, job)) = self.queue.dequeue() {
+            self.pending_queue.push((unit, job));
+            if self.active.len() + self.pending_queue.len() > 1 {
+                jobserver_helper.request_token();
+            }
+        }
+
+        // Do not actually spawn the new work if we've errored out
+        if has_errored {
+            return Ok(());
+        }
+
+        // Now that we've learned of all possible work that we can execute
+        // try to spawn it so long as we've got a jobserver token which says
+        // we're able to perform some parallel work.
+        while self.active.len() < self.tokens.len() + 1 && !self.pending_queue.is_empty() {
+            let (unit, job) = self.pending_queue.remove(0);
+            self.run(&unit, job, cx, scope)?;
+        }
+
+        Ok(())
+    }
+
+    fn grant_rustc_token_requests(&mut self) -> CargoResult<()> {
+        // If we managed to acquire some extra tokens, send them off to a waiting rustc.
+        let extra_tokens = self.tokens.len() - (self.active.len() - 1);
+        for _ in 0..extra_tokens {
+            if !self.to_send_clients.is_empty() {
+                // remove from the front so we grant the token to the oldest
+                // waiter
+                let (id, client) = self.to_send_clients.remove(0);
```

Comment:
Since we're pushing and popping onto this, and it's probably modified relatively frequently, this may actually be a relatively good use for a `VecDeque` perhaps?